src/instructlab/model/full_train.py

```diff
@@ -0,0 +1,313 @@
+        os.path.join(train_args.data_output_dir, "data.jsonl"),
+    )
+
+    # based on the length of the dataset, figure out the max batch len
+    packing_max_batch_len, _ = (
+        multipack_sampler.find_packing_max_batch_len_and_grad_accum(
+            num_gpus=1,
+            avg_sample_len=dataset.get_lengths().mean(),
+            effective_batch_size=train_args.effective_batch_size,
+            max_batch_len_per_gpu=train_args.max_batch_len,
+            is_padding=train_args.is_padding_free,
+            pad_id=tokenizer.pad_token_id,
+            dataset=dataset,
+            seed=47,
+        )
+    )
+
+    collate_fn = partial(pad_collate_fn, pad_token_id=tokenizer.pad_token_id)
+
+    # use a multipack sampler since this is a non-dist. scenario, training library defaults to this as well
+    sampler = multipack_sampler.MultipackDistributedBatchSampler(
+        batch_max_length=packing_max_batch_len,
+        lengths=dataset.get_lengths(),
+        num_replicas=1,
+        rank=0,
+        seed=47,
+        padding=True,
+    )
+    sampler = {"batch_sampler": sampler}
+
+    # 4 workers for the dataloader as compared to the original 8 to optimize performance
+    # this dataloader needs a spawn multiproc method to optimize memory and to work on Apple Silicon.
+    dataloader = DataLoader(
+        dataset,
+        **sampler,
+        num_workers=4,
+        collate_fn=collate_fn,
+    )
+    dataloader.multiprocessing_context = "spawn"
+
+    # set device based off argument given
+    dev = torch.device(device)
+    # auto model based on model path
+    model = AutoModelForCausalLM.from_pretrained(train_args.model_path)
+    if len(tokenizer) > model.config.vocab_size:
+        logger.info(
+            f"WARNING: tokenizer has {len(tokenizer)} tokens but model has {model.config.vocab_size} vocab size"
+        )
+        model.resize_token_embeddings(
+            int(8 * math.ceil(len(tokenizer) / 8.0))
+        )  # make the vocab size multiple of 8 for sharding the embedding layer.
+
+    # Fix any discrepancy between model and tokenizer
+    if (
+        model.config.pad_token_id is not None
+        and tokenizer.pad_token_id is not None
+        and model.config.pad_token_id != tokenizer.pad_token_id
+    ):
+        logger.warning(
+            f"There is a mismatch between pad token id of model ({model.config.pad_token_id}) and tokenizer ({tokenizer.pad_token_id}). Fixing model pad token id to be same as tokenizer's pad token id."
+        )
+        model.config.pad_token_id = tokenizer.pad_token_id
+    if (
+        model.config.bos_token_id is not None
+        and tokenizer.bos_token_id is not None
+        and model.config.bos_token_id != tokenizer.bos_token_id
+    ):
+        logger.warning(
+            f"There is a mismatch between bos token id of model ({model.config.bos_token_id}) and tokenizer ({tokenizer.bos_token_id}). Fixing model bos token id to be same as tokenizer's bos token id."
```

Comment:
A bos token is a beginning of the sequence (or sentence) token, typically used with an eos (end of sequence/sentence) token. But, your point is a good one that this may not be clear to people reading this log message unless they have specific knowledge of tokenizers and model training. This is referring to specific fields in the model's json config versus the tokenizer's json config, which is probably why it refers to them as `bos token id`.

Reply:
I can add some comments here!

Reply:
Thanks, Ben! Yes, agree that some clarifying comments would help

Reply:
added clarification and a comment